internal/pkg/queue/index/manager.go

```diff
@@ -108,39 +131,130 @@ func NewIndexManager(walPath, indexPath, queueDirPath string) (*IndexManager, er
 	}(im, periodicDumpErrChan, periodicDumpStopChan)
 
 	go im.periodicDump(periodicDumpErrChan, periodicDumpStopChan)
+	go im.walCommitsSyncer()
 
 	return im, nil
 }
 
-func (im *IndexManager) Add(host string, id string, position uint64, size uint64) error {
+func (im *IndexManager) unsafeWalSync() error {
+	return im.walFile.Sync()
+}
+
+func (im *IndexManager) walCommitsSyncer() {
+	if swaped := im.walSyncerRunning.CompareAndSwap(false, true); !swaped {
+		panic("walCommitsSyncer already running")
+	}
+	defer im.walSyncerRunning.Store(false)
+
+	lastSyncDuration := time.Duration(0)
+	for {
+		if im.WAL_IO_PERCENT < 1 || im.WAL_IO_PERCENT > 100 {
+			panic(fmt.Errorf("invalid WAL_IO_PERCENT: %d", im.WAL_IO_PERCENT))
+		}
+
+		sleepTime := lastSyncDuration * time.Duration((100-im.WAL_IO_PERCENT)/im.WAL_IO_PERCENT)
+		if sleepTime < im.WAL_MIN_INTERVAL {
+			sleepTime = im.WAL_MIN_INTERVAL
+		}
+		// fmt.Println("lastSyncDuration", lastSyncDuration, "sleepTime", sleepTime)
+		time.Sleep(sleepTime)
+
+		start := time.Now()
+		flyingCommit := im.walCommit.Load()
+		im.Lock()
+		if err := im.unsafeWalSync(); err != nil {
+			slog.Error("failed to sync WAL", "error", err)
+			im.Unlock()
+			return // FIXME: handle this error
```

Comment:
spotted FIXME

Reply:
04e02e60e0002137d6ec0bcf3e81cb081f9778c6

Now it is changed to retry if fsync fails. it's better than losing data.